src/class/dfu/dfu_device.c

```diff
@@ -0,0 +1,648 @@
+};
+
+static tu_lookup_table_t const _dfu_state_table =
+{
+  .count = TU_ARRAY_SIZE(_dfu_state_lookup),
+  .items = _dfu_state_lookup
+};
+
+static tu_lookup_entry_t const _dfu_status_lookup[] =
+{
+  { .key = DFU_STATUS_OK              , .data = "OK" },
+  { .key = DFU_STATUS_ERRTARGET       , .data = "errTARGET" },
+  { .key = DFU_STATUS_ERRFILE         , .data = "errFILE" },
+  { .key = DFU_STATUS_ERRWRITE        , .data = "errWRITE" },
+  { .key = DFU_STATUS_ERRERASE        , .data = "errERASE" },
+  { .key = DFU_STATUS_ERRCHECK_ERASED , .data = "errCHECK_ERASED" },
+  { .key = DFU_STATUS_ERRPROG         , .data = "errPROG" },
+  { .key = DFU_STATUS_ERRVERIFY       , .data = "errVERIFY" },
+  { .key = DFU_STATUS_ERRADDRESS      , .data = "errADDRESS" },
+  { .key = DFU_STATUS_ERRNOTDONE      , .data = "errNOTDONE" },
+  { .key = DFU_STATUS_ERRFIRMWARE     , .data = "errFIRMWARE" },
+  { .key = DFU_STATUS_ERRVENDOR       , .data = "errVENDOR" },
+  { .key = DFU_STATUS_ERRUSBR         , .data = "errUSBR" },
+  { .key = DFU_STATUS_ERRPOR          , .data = "errPOR" },
+  { .key = DFU_STATUS_ERRUNKNOWN      , .data = "errUNKNOWN" },
+  { .key = DFU_STATUS_ERRSTALLEDPKT   , .data = "errSTALLEDPKT" },
+};
+
+static tu_lookup_table_t const _dfu_status_table =
+{
+  .count = TU_ARRAY_SIZE(_dfu_status_lookup),
+  .items = _dfu_status_lookup
+};
+
+#endif
+
+#define dfu_debug_print_context()                                              \
+{                                                                              \
+  TU_LOG2("  DFU at State: %s\r\n         Status: %s\r\n",                     \
+          tu_lookup_find(&_dfu_state_table, _dfu_state_ctx.state),        \
+          tu_lookup_find(&_dfu_status_table, _dfu_state_ctx.status) );    \
+}
+
+//--------------------------------------------------------------------+
+// USBD Driver API
+//--------------------------------------------------------------------+
+void dfu_moded_init(void)
+{
+  _dfu_state_ctx.state = APP_DETACH;    // After init, reset will occur.  We want to be in APP_DETACH to move to DFU_IDLE
+  _dfu_state_ctx.status = DFU_STATUS_OK;
+  _dfu_state_ctx.attrs = 0;
+  _dfu_state_ctx.blk_transfer_in_proc = false;
+
+  dfu_debug_print_context();
+}
+
+void dfu_moded_reset(uint8_t rhport)
+{
+  if (_dfu_state_ctx.state == APP_DETACH)
+  {
+    _dfu_state_ctx.state = DFU_IDLE;
+  } else {
+    switch (_dfu_state_ctx.state)
+    {
+      case DFU_IDLE:
+      case DFU_DNLOAD_SYNC:
+      case DFU_DNBUSY:
+      case DFU_DNLOAD_IDLE:
+      case DFU_MANIFEST_SYNC:
+      case DFU_MANIFEST:
+      case DFU_MANIFEST_WAIT_RESET:
+      case DFU_UPLOAD_IDLE:
+      {
+        _dfu_state_ctx.state = (tud_dfu_firmware_valid_check_cb()) ?  APP_IDLE : DFU_ERROR;
+      }
+      break;
+
+      case DFU_ERROR:
+      default:
+      {
+        _dfu_state_ctx.state = APP_IDLE;
+      }
+      break;
+    }
+  }
+
+  if(_dfu_state_ctx.state == APP_IDLE)
+  {
+    tud_dfu_reboot_to_rt_cb();
```

Comment:
We should remove this API as well, just let the bootloader to decide this by itself. Overall, I think BUS_RESET should just reset the dfu state/variables.

Reply:
Resolved in f830800